compiler/rustc_data_structures/src/sync.rs

```diff
@@ -374,27 +371,25 @@ cfg_if! {
         /// the current thread. Use that for the longest running block.
         #[macro_export]
         macro_rules! parallel {
-            (impl $fblock:tt [$($c:tt,)*] [$block:tt $(, $rest:tt)*]) => {
+            (impl $fblock:block [$($c:expr,)*] [$block:expr $(, $rest:expr)*]) => {
                 parallel!(impl $fblock [$block, $($c,)*] [$($rest),*])
             };
-            (impl $fblock:tt [$($blocks:tt,)*] []) => {
+            (impl $fblock:block [$($blocks:expr,)*] []) => {
                 {
-                    let fblock = IsSend::new(|| $fblock);
+                    let fblock = rustc_data_structures::sync::IsSend(|| $fblock);
+                    let blocks = parallel!(tuple [$($blocks,)*]);
```

Comment:
Thinking about things, we should probably have a Rayon `scope` wrapper which uses `DynSync` and just leave this macro unmodified.